src/gui/sdlmain.cpp

```diff
@@ -2341,9 +2341,15 @@ void GFX_CenterMouse()
 	int current_width  = 0;
 	int current_height = 0;
 
+	auto get_middle = [](const int value) {
+		return lround((value * sdl.desktop.dpi_scale) / 2);
+	};
+
 	assert(sdl.window);
 	SDL_GetWindowSize(sdl.window, &current_width, &current_height);
-	SDL_WarpMouseInWindow(sdl.window, current_width / 2, current_height / 2);
+	SDL_WarpMouseInWindow(sdl.window,
+	                      get_middle(current_width),
+	                      get_middle(current_height));
```

Comment:
`get_canvas_size()` might be a more robust way to get the logical size (and the divide its results by 2).

It's meant to let us avoid having to handle DPI in all the sdlmain code (I think; is that right, @GranMinigun ?).

Reply:
`get_canvas_size()` returns physical pixels, so no, `SDL_GetWindowSize()` is the way to go. But DPI scaling should not be applied here - mouse cursor coordinates are in client points, not physical pixels. This will likely break on macOS, and if what is done here works on Windows, then the correct way to proceed would be reporting the behaviour to the SDL team.

Reply:
My 2c, I don't think you can ever be totally oblivious to DPI in the code, but that's not a problem. We just need to make that always explicit whether an argument/variable is a physical (actual physical pixel) coordinate or a logical (client) coordinate. And generally you should default to client coordinates, so if we are consistent about that, marking physical coords explicitly in the namings should de sufficient.

Otherwise we get into a mess like the weird issue I stumbled into the other day. I was really exhausted then so maybe I drew the wrong conclusions, but I was dead certain then that some functions get repeatedly called with both logical and physical coordinates, and that's how the bug manifests...
https://github.com/dosbox-staging/dosbox-staging/issues/2603

Reply:
> This will likely break on macOS, and if what is done here works on Windows, then the correct way to proceed would be reporting the behaviour to the SDL team.

I agree with that, just pointing out we can ifdef it to not apply the division on Apple as a temp workaround. I'll test it soon.

Reply:
Relaying from @GranMinigun :

> for now this workaround can be `#ifdef`ed (to Windows) to use `get_canvas_size()` on Windows.

Reply:
Added `#ifdef` to use `get_canvas_size` on Windows.

Reply:
Or, maybe the best solution would be to use SDL_GetRendererOutputSize? https://wiki.libsdl.org/SDL2/SDL_GetRendererOutputSize

Sorry, this is all very confusing to me. Why does anyone assume I need window size in some strange units? I just want size in pixels, sometimes a DPI (horizontal and vertical - I assume they might be different).

Reply:
Because one may have displays with different scaling factors applied, and the window is supposed to keep its relative size the same on both in such cases. Cursor should also preserve its speed, so that there's no sudden jump in sensitivity by merely moving it between displays. One display may be 4K, another Full HD, both may have similar physical size. Hence the need for DPI scaled coordinates (which Apple calls "client points"). HiDPI shenanigans, yes.

`SDL_GetRendererOutputSize()`, as hinted by its name, works only with SDL's renderer (`texture` output in DOSBox). It's wrapped by `get_canvas_size()` function.

Reply:
Cursor speed can depend on resolution and DPI. For the GUI scaling - when I was young, the system font size was used for this, it worked. I find the desktop scaling useless, it makes everything blurry.

Unfortunately, I don't know whether it is the `SDL_GetWindowSize` which is not consistent, or `SDL_WarpMouseInWindow`, or is it due to some capability flag in Windows EXE file - I only use Linux at home.